src/tasks/publish/index.ts

```diff
@@ -0,0 +1,226 @@
+      ensAddress: networkConfig.ensAddress
+    }
+  )
+
+  const prevVersion = await _getLastestVersionIfExists(appName, provider)
+
+  const { bump, nextVersion } = parseAndValidateBumpOrVersion(
+    bumpOrVersion,
+    prevVersion ? prevVersion.version : undefined
+  )
+  logMain(`Applying version bump ${bump}, next version: ${nextVersion}`)
+
+  // Using let + if {} block instead of a ternary operator
+  // to assign value and log status to console
+  let contractAddress: string
+  if (onlyContent) {
+    logMain('No contract used for this version')
+    contractAddress = zeroAddress
+  } else if (existingContractAddress) {
+    logMain('Using provided contract address')
+    contractAddress = existingContractAddress
+  } else if (!prevVersion || bump === 'major') {
+    logMain('Deploying new implementation contract')
+    contractAddress = await _deployMainContract(contractName, noVerify, bre)
+  } else {
+    logMain('Reusing contract from previous version')
+    contractAddress = prevVersion.contractAddress
+  }
+  logMain(`contractAddress: ${contractAddress}`)
+
+  // Prepare release directory
+  // npm run build must create: index.html, src.js, script.js
+  logMain(`Building app front-end at ${appSrcPath}`)
+  await execa('npm', ['run', 'build'], { cwd: appSrcPath })
+
+  // Generate and validate Aragon artifacts, release files
+  logMain(`Generating Aragon app artifacts`)
+  await generateArtifacts(distPath, bre)
+  validateArtifacts(distPath)
+
+  // Upload release directory to IPFS
+  logMain('Uploading release assets to IPFS...')
+  const contentHash = await uploadReleaseToIpfs(distPath, {
+    ipfsProvider,
+    ignorePatterns: ipfsIgnore
+  })
+  logMain(`Release assets uploaded to IPFS: ${contentHash}`)
+
+  // Generate tx to publish new app to aragonPM
+  const versionInfo = {
+    version: nextVersion,
+    contractAddress,
+    contentUri: apm.toContentUri('ipfs', contentHash)
+  }
+
+  const txData = await apm.publishVersion(appName, versionInfo, provider, {
+    managerAddress
+  })
+
+  logMain(
+    `Successfully generated TX data for publishing ${appName} version ${nextVersion}
+
+to: ${txData.to}
+data: ${apm.encodePublishVersionTxData(txData)}
+`
+  )
+}
+
+/**
+ * Returns latest version given a partial or full appName
+ * Returns undefined if repo does not exists
+ * @param appName "finance" | "finance.aragonpm.eth"
+ * @param provider
+ */
+async function _getLastestVersionIfExists(
+  appName: string,
+  provider: ethers.providers.Provider
+): Promise<apm.ApmVersion | undefined> {
+  const fullAppName = getFullAppName(appName)
+  // Check ENS name first since ethers causes an UnhandledPromiseRejectionWarning
+  const repoAddress = await provider.resolveName(fullAppName)
+  if (!repoAddress) return
+
+  // Check for latest version but expect errors
+  try {
+    return await apm.getRepoVersion(repoAddress, 'latest', provider)
+  } catch (e) {
+    if (e.message.includes('ENS name not configured')) return
```

Comment:
We shouldn't get this since we've already checked that the ENS name, right?

Reply:
True! Good pick